### PR TITLE
Fix type operator post-increment, shift and unsigned types

### PR DESCRIPTION
**Pull request by xezon**

### CODE REVIEW

File: deps/baseconfig/src/typeoperators.h [hunk 1]
```diff
@@ -35,20 +35,21 @@
...
+__forceinline constexpr ENUMTYPE operator^(ENUMTYPE const a, ENUMTYPE const b) { return ENUMTYPE(((UNDERLYING)a) ^ ((UNDERLYING)b)); } \
+__forceinline ENUMTYPE &operator^=(ENUMTYPE &a, ENUMTYPE const &b) { return (ENUMTYPE &)(((UNDERLYING &)a) ^= ((UNDERLYING &)b)); } \
+__forceinline ENUMTYPE &operator&=(ENUMTYPE &a, ENUMTYPE const &b) { return (ENUMTYPE &)(((UNDERLYING &)a) &= ((UNDERLYING &)b)); } \
+__forceinline ENUMTYPE &operator|=(ENUMTYPE &a, ENUMTYPE const &b) { return (ENUMTYPE &)(((UNDERLYING &)a) |= ((UNDERLYING &)b)); } \
+__forceinline constexpr ENUMTYPE operator<<(ENUMTYPE a, size_t const b) { return (ENUMTYPE)(((UnsignedInteger<ENUMTYPE>::type)a) << b); } \
+__forceinline constexpr ENUMTYPE operator>>(ENUMTYPE a, size_t const b) { return (ENUMTYPE)(((UnsignedInteger<ENUMTYPE>::type)a) >> b); } \
+__forceinline ENUMTYPE &operator<<=(ENUMTYPE &a, size_t const b) { return (ENUMTYPE &)((UnsignedInteger<ENUMTYPE>::type &)a <<= b); } \
+__forceinline ENUMTYPE &operator>>=(ENUMTYPE &a, size_t const b) { return (ENUMTYPE &)((UnsignedInteger<ENUMTYPE>::type &)a >>= b); } \
```
Comment (xezon) [inline deps/baseconfig/src/typeoperators.h]: Now shifting is always treated as unsigned type. We could treat it as `UNDERLYING`, but I don't know if that makes sense. I think `UnsignedInteger<ENUMTYPE>` is safe for now.

File: deps/baseconfig/src/typeoperators.h [hunk 2]
```diff
@@ -35,20 +35,21 @@
 #ifdef __cplusplus
 #if !defined(DEFINE_ENUMERATION_BITWISE_OPERATORS)
 // clang-format off
-#define DEFINE_ENUMERATION_BITWISE_OPERATORS(ENUMTYPE) \
+#define DEFINE_ENUMERATION_BITWISE_OPERATORS_T(ENUMTYPE, UNDERLYING) \
```
Comment (xezon) [inline deps/baseconfig/src/typeoperators.h:38]: Macros now accept second argument for underlying type when so required.

File: deps/baseconfig/src/typeoperators.h [hunk 3]
```diff
@@ -35,20 +35,21 @@
...
+__forceinline ENUMTYPE &operator^=(ENUMTYPE &a, ENUMTYPE const &b) { return (ENUMTYPE &)(((UNDERLYING &)a) ^= ((UNDERLYING &)b)); } \
+__forceinline ENUMTYPE &operator&=(ENUMTYPE &a, ENUMTYPE const &b) { return (ENUMTYPE &)(((UNDERLYING &)a) &= ((UNDERLYING &)b)); } \
+__forceinline ENUMTYPE &operator|=(ENUMTYPE &a, ENUMTYPE const &b) { return (ENUMTYPE &)(((UNDERLYING &)a) |= ((UNDERLYING &)b)); } \
+__forceinline constexpr ENUMTYPE operator<<(ENUMTYPE a, size_t const b) { return (ENUMTYPE)(((UnsignedInteger<ENUMTYPE>::type)a) << b); } \
+__forceinline constexpr ENUMTYPE operator>>(ENUMTYPE a, size_t const b) { return (ENUMTYPE)(((UnsignedInteger<ENUMTYPE>::type)a) >> b); } \
+__forceinline ENUMTYPE &operator<<=(ENUMTYPE &a, size_t const b) { return (ENUMTYPE &)((UnsignedInteger<ENUMTYPE>::type &)a <<= b); } \
+__forceinline ENUMTYPE &operator>>=(ENUMTYPE &a, size_t const b) { return (ENUMTYPE &)((UnsignedInteger<ENUMTYPE>::type &)a >>= b); } \
 }
+#define DEFINE_ENUMERATION_BITWISE_OPERATORS(ENUMTYPE) DEFINE_ENUMERATION_BITWISE_OPERATORS_T(ENUMTYPE, SignedInteger<ENUMTYPE>::type)
```
Comment (xezon) [inline deps/baseconfig/src/typeoperators.h:52]: Standard macro assumes signed integer.

File: deps/baseconfig/src/typeoperators.h [hunk 4]
```diff
@@ -71,27 +72,28 @@ __forceinline ENUMTYPE &operator>>=(ENUMTYPE &a, SizedInteger<ENUMTYPE>::type co
...
+__forceinline constexpr ENUMTYPE operator-(ENUMTYPE const a, ENUMTYPE const b) { return (ENUMTYPE)(((UNDERLYING)a) - ((UNDERLYING)b)); } \
+__forceinline constexpr ENUMTYPE operator*(ENUMTYPE const a, ENUMTYPE const b) { return (ENUMTYPE)(((UNDERLYING)a) * ((UNDERLYING)b)); } \
+__forceinline constexpr ENUMTYPE operator/(ENUMTYPE const a, ENUMTYPE const b) { return (ENUMTYPE)(((UNDERLYING)a) / ((UNDERLYING)b)); } \
+__forceinline constexpr ENUMTYPE operator%(ENUMTYPE const a, ENUMTYPE const b) { return (ENUMTYPE)(((UNDERLYING)a) % ((UNDERLYING)b)); } \
+__forceinline ENUMTYPE &operator+=(ENUMTYPE &a, ENUMTYPE const b) { return (ENUMTYPE &)((UNDERLYING &)a = ((UNDERLYING &)a) + ((ENUMTYPE)b)); } \
+__forceinline ENUMTYPE &operator-=(ENUMTYPE &a, ENUMTYPE const b) { return (ENUMTYPE &)((UNDERLYING &)a = ((UNDERLYING &)a) - ((ENUMTYPE)b)); } \
+__forceinline ENUMTYPE operator++(ENUMTYPE const &a, int) { return (ENUMTYPE)(((UNDERLYING &)a)++); } \
+__forceinline ENUMTYPE operator--(ENUMTYPE const &a, int) { return (ENUMTYPE)(((UNDERLYING &)a)--); } \
```
Comment (xezon) [inline deps/baseconfig/src/typeoperators.h]: This is post-increment. It was implemented as pre-increment before and was therefore wrong.

**Comment by duncanspumpkin** [comment]
If you make this all constexpr we can add tests that are checked at compile time for types I think. `static_assert(NameKey(1)++ == NameKey(1));`, `static_assert(++NameKey(1) == NameKey(2));`

**Comment by xezon** [comment]
Let me check.

**Comment by xezon** [comment]
Ok. That works.

```c++
enum ea
{
    ea0 = 0,
    ea1,
    ea2,
    ea3,
    ea4,
};

enum eb
{
    eb0 = 0,
    eb1 = 1,
    eb2 = 2,
    eb4 = 4,
    eb8 = 8,
};

DEFINE_ENUMERATION_OPERATORS(ea)
DEFINE_ENUMERATION_BITWISE_OPERATORS(eb)

static_assert(++ea(ea0) == ea(ea1));
static_assert(ea(ea0)++ == ea(ea0));
static_assert(eb(eb1) << 1u == eb(eb2));
static_assert((eb(eb2) | eb(eb4)) == eb(eb2) + eb(eb4));
```

**Review by OmniBlade — APPROVED** [review]

File: deps/baseconfig/src/typesize.h [hunk 5]
```diff
@@ -20,33 +20,26 @@
...
+
+template<> struct UnsignedIntegerForSize<1>{ typedef uint8_t  type; };
+template<> struct UnsignedIntegerForSize<2>{ typedef uint16_t type; };
+template<> struct UnsignedIntegerForSize<4>{ typedef uint32_t type; };
+template<> struct UnsignedIntegerForSize<8>{ typedef uint64_t type; };
+
+template<class T> struct SignedInteger{ typedef typename SignedIntegerForSize<sizeof(T)>::type type; };
+template<class T> struct UnsignedInteger{ typedef typename UnsignedIntegerForSize<sizeof(T)>::type type; };
+
+// clang-format off
```
Comment (xezon) [inline deps/baseconfig/src/typesize.h]: Let me fix this before merge.

Comment (xezon) [reply deps/baseconfig/src/typesize.h]: Fixed.

**Pull request merged**